eo-maven-plugin/src/main/java/org/eolang/maven/OptimizeMojo.java

```diff
@@ -138,18 +142,150 @@ private int optimized(final ForeignTojo tojo, final Optimization optimization)
      * Common optimization for all tojos.
      * @return Optimization for all tojos.
      */
-    private Optimization optimization() {
-        final Optimization opt;
+    private Function<XML, XML> optimization() {
+        final Function<XML, XML> opt;
         final Train<Shift> train = this.measured(new TrParsing());
         if (this.trackOptimizationSteps) {
-            opt = new OptSpy(
+            opt = new OptimizeMojo.OptSpy(
                 train,
                 this.targetDir.toPath().resolve(OptimizeMojo.STEPS)
             );
         } else {
-            opt = new OptTrain(train);
+            opt = new OptimizeMojo.OptTrain(train);
         }
         return opt;
     }
 
+    /**
+     * Optimization that spies.
+     * @since 0.68.0
+     */
+    public static final class OptSpy implements Function<XML, XML> {
+        /**
+         * Optimizations train.
+         */
+        private final Train<Shift> train;
+
+        /**
+         * Where to track optimization steps.
+         */
+        private final Path target;
+
+        /**
+         * Ctor.
+         * @param target Where to track optimization steps.
+         */
+        public OptSpy(final Path target) {
+            this(OptTrain.DEFAULT_TRAIN, target);
+        }
+
+        /**
+         * The main constructor.
+         * @param trn Optimizations train.
+         * @param target Where to track optimization steps.
+         */
+        public OptSpy(final Train<Shift> trn, final Path target) {
+            this.train = trn;
+            this.target = target;
+        }
+
+        @Override
+        public XML apply(final XML xml) {
+            final Path dir = new Place(xml.xpath("/program/@name").get(0)).make(this.target, "");
+            Logger.debug(this, "Optimization steps will be tracked to %[file]s", dir);
+            return new OptTrain(new SpyTrain(this.train, dir)).apply(xml);
+        }
+    }
+
+    /**
+     * Optimisation train of XLS`s.
+     * @since 0.68.0
+     * @todo #3115:30min Return constant-folding.xsl when it's ready.
+     * This optimization was removed from the train because it's not really
+     * ready and works only with `bool` object which was removed. We
+     * need to make this optimization great again and add to the train.
+     */
+    public static final class OptTrain implements Function<XML, XML> {
```

Comment:
@ArtemGet can this class be package private?

Reply:
Yes, because used only inside package.